store/multi_writer_app_store.go

```diff
@@ -354,3 +354,16 @@ func (s *multiWriterStoreSnapshot) Range(prefix []byte) plugin.RangeData {
 
 	return ret
 }
+
+func (m *MultiWriterAppStore) VersionExists(version int64) bool {
+	return m.appStore.VersionExists(version)
+}
+
+func (m *MultiWriterAppStore) RetrieveVersion(version int64) (VersionedKVStore, error) {
+	reader, err := newMultiWriterVersionReader(*m, version)
```

Comment:
I don't think we need all these new stores (`iAVLVersionReader`, `multiWriterVersionReader`), you should be able to do something very similar to what `GetSnapshot()` does, in fact rather than adding new methods to `VersionedKVStore` you can probably modify `GetSnapshot()` -> `GetSnapshot(version int64)` and use `version == 0` to retrieve the last snapshot, and `version > 0` to retrieve a specific version.

`splitStore` is still going to be required to allow in-memory writes.

Reply:
You need `value, _, _ := s.IAVLStore.tree.GetVersionedWithProof(key, s.version)` somehwer as in
```to
func (s *SomeKVReader) Get(key []byte) []byte {
	value, _, _ := s.IAVLStore.tree.GetVersionedWithProof(key, s.version)
	return value
}
```

Reply:
You need to obtain the `ImmutableTree` at the desired height/version and then use the regular Get/Has/Range like the snapshots do. This is basically was `MutableTree.GetVersionedWithProof` is doing every time you call it. Similarly snapshots work off the last saved `ImmutableTree`, and now we need to extend them to work with any previously saved `ImmutableTree` version.

Reply:
I have renamed iAVLVersionReader to a Snapshot and merged multiWriterVersionReader and newMultiWriterStoreSnapshot.